url: update `filterURLForDisplay` to include all image, video, and audio file types

brookewp opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Closes #53154 

When replacing a file that is not a jpg, jpeg, gif, png, or svg, the link in the 'Replace' `Popover` overflows the width and covers up the edit icon. It makes editing extremely difficult as the link is much easier to click than the icon. 

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

The current RegEx used only captures limited image file types. However, this filter is used in `LinkControl`, which is used in many other cases, including the Video and Audio blocks, where it is completely broken. 

This particular RegEx is for the URL's display, so I'm unsure if there is a need for it to be so specific. If so, that might be beyond this PR's scope, but I wanted to see if we could get something started here, as it is a frustrating UX. 

## How?

By changing the RegEx for media in `filterURLForDisplay` to capture any file extension. 

## Testing Instructions

Ensure tests still pass: `npm run test:unit packages/url/src/test/index.js`

### For MediaReplaceFlow

1. Add a block that uses `LinkControl` (i.e. Video, Audio, Image)
2. Add a file format that wasn't included before (jpg, jpeg, gif, png, or svg)
3. See the URL is shortened, and the edit icon can easily be clicked again 

| Before  | After |
| ------------- | ------------- |
|  <img width="352" alt="Screenshot 2023-09-28 at 1 55 17 PM" src="https://github.com/WordPress/gutenberg/assets/35543432/8b52cd4d-cd8f-4efa-a2eb-ba802caa8fb4"> | <img width="351" alt="Screenshot 2023-09-28 at 1 53 48 PM" src="https://github.com/WordPress/gutenberg/assets/35543432/1e775bbc-b3c3-49c3-9bf6-863c50dbf588">  |

### For LinkPreview

Ensure tests still pass: `npm run test:unit packages/block-editor/src/components/link-control/test/index.js`

1. Add a Paragraph block 
2. Add an inline link and open the UI to edit
3. Hover on or tab to the link to see a tooltip with the full URL

<img width="426" alt="Screenshot 2023-10-06 at 9 04 45 AM" src="https://github.com/WordPress/gutenberg/assets/35543432/53b5c88f-7442-4f37-a287-8d62cc0d2146">

brookewp committed:
url: update display filter to include all file types

getdave reviewed (commented):
Thanks for the PR 👍 

I wonder whether you would be open to adding some unit tests for this function? Within those we could assert on your change.

Then future changes to this function will be easier to review and it will help to avoid any regressions.

Thanks.

tyxla reviewed (commented):
This is a neat fix, thanks 🙌 

I wonder if instead, it's better to expand the list of allowed file types to contain the ones that are currently missing. It is [a limited set of file extensions](https://codex.wordpress.org/Uploading_Files), after all (even if it's an [extensible list](https://developer.wordpress.org/reference/hooks/upload_mimes/)). That would definitely make the regex longer, so don't consider that a requirement, but rather something to consider.

Otherwise, I agree that this change could use some tests to ensure we're getting the right results. There are already tests that cover cases like multiple periods, etc., but I guess we can add some tests that cover additional formats that weren't supported before.

t-hamano reviewed (commented):
Thanks for the PR, @brookewp!

As I understand it, the purpose of this function is to ensure that when a URL points to a file, the three-letter file name and extension are displayed at the end. I think this is to make it easier to understand what file is linked to instead of the URL starting from the beginning.

Therefore, I personally think it makes sense to target all files, not just media files.

For example, the inline link allows you to specify URLs with any extension that are not allowed to be uploaded by default. Below is an example where a `.diff` file with no media upload allowed was entered as a URL:

| trunk | this PR |
|--------|--------|
| ![before](https://github.com/WordPress/gutenberg/assets/54422211/af25b37a-0ba0-4962-ab3b-30e0c1e0a54c)| ![after](https://github.com/WordPress/gutenberg/assets/54422211/3a48b4a3-5183-4407-a89e-4f784a37bcf4) | 

If it makes sense to allow all extensions, you might be better off with more explicit code like this:

```javascript
const isFileUrl = !! filteredURL.match( /\/([^\/?]+)\.(?:[\w]+)(?=\?|$)/ );

if ( ! maxLength || filteredURL.length <= maxLength || ! isFileUrl ) {
	return filteredURL;
}
```

What do you think, everyone?

brookewp committed:
Restore regex with defined media files with extended list

brookewp commented:
Thanks all for the feedback!

@t-hamano, if I'm understanding correctly, you're suggesting this behavior for all file URLs, for visual consistency. Perhaps [that is something](https://github.com/WordPress/gutenberg/pull/54920#pullrequestreview-1650680090) @jasmussen could weigh in on? 

In the meantime, I'm happy to have an explicit list from the WP docs that @tyxla shared so we can safely and quickly resolve the UI issue for our UX. In that case, would other tests for various media types be helpful? Many use jpg, so we could consider replacing some with varying formats.

brookewp commented on packages/url/src/filter-url-for-display.js:
```diff
@@ -21,8 +21,8 @@ export function filterURLForDisplay( url, maxLength = null ) {
-	const mediaRegexp = /([\w|:])*\.(?:jpg|jpeg|gif|png|svg)/;
-
+	const mediaRegexp =
+		/([\w|:])*\.(?:jpg|jpeg|gif|png|svg|ico|webp|mp3|m4a|ogg|wav|mp4|m4v|mov|wmv|avi|mpg|ogv|3gp|3g2)/;
```
I didn't include [the documents](https://codex.wordpress.org/Uploading_Files) in the accepted file types, as this isn't a problem with the File block. But we could always add those to future-proof a bit.

tyxla replied:
Feel free to add them all IMHO.

tyxla reviewed (approved):
This LGTM 👍 

I'd be happy with a solution that works with all files too, since as I mentioned in the earlier review, the list of allowed files can be expanded by a filter.

However, this seems to resolve the aforementioned issue for the commonly supported files, so I'm good with it.

🚀

jasmussen commented:
> Perhaps https://github.com/WordPress/gutenberg/pull/54920#pullrequestreview-1650680090 @jasmussen could weigh in on?

I share that instinct, and would love to see this across. It isn't useful to see a very very very long URL like that. If we absolutely have to see the full URL, it could be a tooltip, or you could see it in the input field when editing.

getdave commented:
> > Perhaps [#54920 (review)](https://github.com/WordPress/gutenberg/pull/54920#pullrequestreview-1650680090) @jasmussen could weigh in on?
> 
> It isn't useful to see a very very very long URL like that. If we absolutely have to see the full URL, it could be a tooltip, or you could see it in the input field when editing.

I would like to sound a note of caution here. Based on my experience with how users utilise `<LinkControl>` there have been multiple points where users have wanted to see the entire URL. 

Why?

Because sometimes this is the best way to help differentiate between near identical files. In the past we did try truncating the URL to a single line and we received near immediate feedback that this was causing severe difficulty for a set of our users.

This will continue to be a problem until we provide a better way to handle selecting attachments when creating links (which is another outstanding LinkControl issue 😓). A good example of this is PDF files.

Therefore, whilst I don't disagree that 80% of the time seeing a shortened URL is preferable, I would say that for attachments (at least) we should provide some means to see the _entire_ URL. 

Note that whatever solution is chosen needs to account for non-visual users and must be accessible and perceivable by users of a screenreader/keyboard.


🙏 Please note that we should also be testing this on the standard Link UI that is used when creating links in the paragraph block. That looks different to the Media block usage and needs to be considered. Thank you in advance 🙇

getdave reviewed (changes requested):
After some consideration I decided I feel strongly enough about [this](https://github.com/WordPress/gutenberg/pull/54920#issuecomment-1746384383) to leave a blocking review.

This will avoid us shipping a experience which could be detrimental to some users.

Thank you 🙇

brookewp commented:
Thanks all for the input! And @getdave for your caution, as I know you've been working on LinkControl a lot, so I appreciate that you've shared your experience with user feedback. 

> Please note that we should also be testing this on the standard Link UI that is used when creating links in the paragraph block. 

If I'm looking at the correct Link UI, then the behavior already exists in trunk for limited image file types: 

<img width="540" alt="Screenshot 2023-10-04 at 7 45 23 PM" src="https://github.com/WordPress/gutenberg/assets/35543432/9d422186-cf00-4943-b4e0-283ca119b27d">

>This will avoid us shipping a experience which could be detrimental to some users.

I didn't include PDFs and other documents, just video, image, and audio files, to try to resolve the UI issue with these blocks so it isn't so difficult to replace media. Since we have this in place for image files already, it didn't appear to be a big change to extend it to include a few other image formats, video, and audio files. 

I do understand the hesitation in adding more files or all of them. This is why I decided to limit the scope of the files after the initial feedback, to make a smaller and more cautious change. If this is still too big and we need to reassess the design and implementation, it might be beyond the scope of this PR.

getdave commented:
No problem. Appreciate you being open to feedback.

I'm keen not to overcomplicate this but I am also keen not to regress something that users actively asked us _not_ to do.

Therefore I wonder if one solution could be:

- add a tooltip on hover which could display the full URL.
- add an `aria-describedby` or [`aria-details`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-details) attribute to the element containing the url to make it perceivable by assitive tech (cc @alexstine for confirmation this is appropriate). 

>  the behavior already exists in trunk for limited image file types:

I think that's because (on anything other than `localhost`) there would be a preview of the media file if it is an _image_ in the "rich preview" section (once the link is submitted).

alexstine commented:
@getdave I'm not entirely sure how to test this... Not really following this well.

getdave commented:
> @getdave I'm not entirely sure how to test this... Not really following this well.

Sorry. Here is some more context.

This PR seeks to truncate the `url` that is displayed in the Link UI's "preview". For a normal link that URL would be something like `www.wordpress.org`. However, if it's an attachment (e.g. media) then it can be a very long string (think large filenames...etc) and so we want to truncate the text but show the final portion of the `url`.

As an example the URL `www.mysite.com/uploads/images/2019/09/my-really-long-image-file-path-is-super-long-and-hard-to-type-out-haha.jpg` would become something like `../../hard-to-type-out-haha.jpg`.

Note that this truncation is _not_ a visual thing. It actually removes the characters from the string.

I raised a red flag that (when we tried a similar thing in the past) users communicated that they needed to see the entire URL in certain circumstances as a way of differentiating between attachment files (think PDF files).

My suggestion is:

- allow truncation of the URL for visual / aeathetic purposes
- provide another means to view/access the _full_ URL

It is for the later point that I"m requesting your input on the approach. Hypothetically would it be ok to do something like this where the element referenced as `aria-describedby` contains the _full_ URL and also acts as a tooltip for sighted users:

```html
<span aria-describedby="full-url">../../hard-to-type-out-haha.jpg</span>
<span class="tooltip" id="full-url">www.mysite.com/uploads/images/2019/09/my-really-long-image-file-path-is-super-long-and-hard-to-type-out-haha.jpg</span>
```

I'm not sure how easy it is for you to access the HTML example above so if that's not helpful then perhaps @brookewp would consider pulling together an working example implementation in this PR for review?

Thanks in advance.

alexstine commented:
@getdave I don't think `aria-describedby` will work on a span as its not an interactive element.

getdave commented:
> @getdave I don't think `aria-describedby` will work on a span as its not an interactive element.

Ok thanks.

Interestingly I just noticed that [our `<Tooltip>` component](https://wordpress.github.io/gutenberg/?path=/docs/components-tooltip--docs) adds an `aria-describedby` (referencing the tooltip popover) to whatever triggers/acts as anchor for the tooltip. 

Note this from the docs:

> The text shown in the tooltip when anchor element is focused or hovered.

Given the above, and Alex's reminder to use "interactive" elements, then perhaps we can anchor the tooltip to [the `<a>` tag in the Link Preview](https://github.com/WordPress/gutenberg/blob/fcb107978ef8b06ef1996176c46fada3e0b38eec/packages/block-editor/src/components/link-control/link-preview.js#L90-L95) instead?

https://github.com/WordPress/gutenberg/blob/fcb107978ef8b06ef1996176c46fada3e0b38eec/packages/block-editor/src/components/link-control/link-preview.js#L90-L95

Seems like that would be accessible (in all senses of the word) by default given our `Tooltip` implementation, but as with all such things it would require testing once implemented to be sure.

alexstine commented:
@getdave Would not hurt to try.

richtabor commented:
We should certainly come to a solution here; currently it's a bit of a mess — you can't click the "Edit" link as-is. 

![CleanShot 2023-10-05 at 18 49 18](https://github.com/WordPress/gutenberg/assets/1813435/bfc1a936-474c-439a-b92e-bcd93095e8c2)

brookewp committed:
Add tooltip to LinkPreview and remove from MediaReplaceFlow

joedolson commented:
Since the link URL that's displayed is a link, it should work just fine to use the Tooltip component to display the full URL. It wouldn't be different from other usages of a Tooltip from an accessibility perspective. That also means that the information would be presented in a manner consistent with other areas of the editor, and would inherit any future improvements we make to the component, so I think that's a good strategy.

I'm removing the needs accessibility feedback label on this; you can add it back when there's a PR with this change to review.

brookewp committed:
Restore regex to display all file urls

brookewp committed:
Add test to LinkPreview to check for tooltip

brookewp committed:
Add test for filter to verify newly included formats

brookewp commented:
I've added a Tooltip to LinkPreview which required a couple other changes:
- The tooltip from MediaReplaceFlow can be removed so the tooltip isn't duplicated from LinkPreview
- The filter regex was restored to include all files again, so we don't have full URLs and tooltips for only some urls. 

I also added a couple of tests. One to check the tooltip and one for new formats in the filter, based on what @tyxla suggested:
>we can add some tests that cover additional formats that weren't supported before

getdave commented on packages/url/src/test/index.js:
```diff
@@ -273,6 +273,21 @@ describe( 'getFilename', () => {
 		expect( getFilename( 'image.jpg' ) ).toBe( 'image.jpg' );
 	} );
 
+	it( 'returns the filename for a variety of formats', () => {
+		expect( getFilename( 'https://wordpress.org/file.pdf' ) ).toBe(
+			'file.pdf'
+		);
+		expect(
+			getFilename( 'https://wordpress.org/image.webp?query=test' )
+		).toBe( 'image.webp' );
+		expect( getFilename( 'https://wordpress.org/video.mov#anchor' ) ).toBe(
+			'video.mov'
+		);
+		expect(
+			getFilename( 'http://localhost:8080/a/path/to/audio.mp3' )
+		).toBe( 'audio.mp3' );
+	} );
+
```
It feels like we could make a single test by combining this one with `returns the filename part of the URL` and using `it.each()` to run over a variety of formats. What do you think?

tyxla replied:
I wouldn't be opposed to that, since it could add some extra durability to this feature in the case of future changes.

brookewp replied:
Sounds good!

brookewp replied:
Do we need `it.each()` in this case? Rather than just combining them? 

https://github.com/WordPress/gutenberg/blob/d8bc25dddbf23e05bc2a0b4339a13789ec11777c/packages/url/src/test/index.js#L257-L286

andrewhayward replied:
Strictly speaking, we don't _need_ to use `each()` here, but this is precisely the sort of scenario that it was designed for, and can make the intent of the test much clearer.

```js
it.each( [
  [ 'https://wordpress.org/image.jpg', 'image.jpg' ],
  [ 'https://wordpress.org/image.jpg?query=test', 'image.jpg' ],
  [ 'https://wordpress.org/image.jpg#anchor', 'image.jpg' ],
  [ 'http://localhost:8080/a/path/to/an/image.jpg', 'image.jpg' ],
  [ '/path/to/an/image.jpg', 'image.jpg' ],
  [ 'path/to/an/image.jpg', 'image.jpg' ],
  [ '/image.jpg', 'image.jpg' ],
  [ 'https://wordpress.org/file.pdf', 'file.pdf' ],
  [ 'https://wordpress.org/image.webp?query=test', 'image.webp' ],
  [ 'https://wordpress.org/video.mov#anchor', 'video.mov' ],
  [ 'http://localhost:8080/a/path/to/audio.mp3', 'audio.mp3' ],
] )( 'returns the filename part of the URL: %s', ( url, filename ) => { 
  expect( getFilename( url ) ).toBe( filename ); 
} ); 
```

Rather than having to read through the entire test to understand what's expected, we can see quite quickly that the intention is to have the same expectation for each set of parameters.

If we need to test additional parameters, it is clearer how to do that. And if we need to change the expectation, we will have more certainty that we've correctly applied that in every case.

brookewp replied:
Thank you, @andrewhayward 🙌 This is very helpful!

getdave replied:
Exactly Andrew. This is precisely what I had in mind. Thank you.

getdave commented on packages/block-editor/src/components/link-control/test/index.js:
```diff
@@ -2056,6 +2056,26 @@ describe( 'Addition Settings UI', () => {
+
+	it( 'should show tooltip with full URL alongside filtered display', async () => {
+		const user = userEvent.setup();
+		const url = 'https://example.com';
```
I think our test should be for a longer URL and probably one that includes a file format. That would add an extra layer of resilience to the test.

tyxla replied:
Agreed, a longer URL would make sense here.

brookewp replied:
Will do!

getdave reviewed (approved):
Thank you for all the iterations here. I tested this and it works as described.

I tested standard Link UI and also the one in Media replace flow.

[Example PDF.pdf](https://github.com/WordPress/gutenberg/files/12843905/Example.PDF.pdf)

I left some minor suggestions.

Great work 👏

getdave commented:
> ...when there's a PR with this change to review.

Re-adding the a11y label now as requested.

t-hamano reviewed (approved):
I think this latest PR looks like it provides a good way to consistently see the full URL in both the inline link and MediaReplaceFlow.

The screencast below shows that you can display a tooltip with the full URL using only the keyboard, and that a screen reader (NVDA) will also read the full URL.

https://github.com/WordPress/gutenberg/assets/54422211/2b7bc9a9-890a-4cc3-91dd-4b72117ed4be

tyxla commented on packages/block-editor/src/components/link-control/test/index.js:
```diff
@@ -2056,6 +2056,26 @@ describe( 'Addition Settings UI', () => {
+
+		const link = screen.getByRole( 'link' );
+
+		expect( screen.getByRole( 'link' ) ).toHaveTextContent(
```
We can totally go with the constant above instead of doing another query here. 

```suggestion
		expect( link ).toHaveTextContent(
```

tyxla reviewed (approved):
Nice work @brookewp 🙌 🚀 Thanks!

andrewhayward commented on packages/url/src/filter-url-for-display.js:
```diff
@@ -21,7 +21,7 @@ export function filterURLForDisplay( url, maxLength = null ) {
 	}
 
-	const mediaRegexp = /([\w|:])*\.(?:jpg|jpeg|gif|png|svg)/;
+	const mediaRegexp = /\/([^\/?]+)\.(?:[\w]+)(?=\?|$)/;
```
Small point, but I wonder if it's worth renaming this regex now that it's not explicitly looking for "media" any more?

brookewp replied:
Good point!

brookewp committed:
Update link control test based on feedback

brookewp committed:
Combine new test with existing filename test

brookewp committed:
Update const naming and add comment

brookewp committed:
Clean up test by replacing with it.each

tyxla commented on packages/block-editor/src/components/link-control/test/index.js:
```diff
@@ -2056,6 +2056,25 @@ describe( 'Addition Settings UI', () => {
+
+		const link = screen.getByRole( 'link' );
+
+		expect( link ).toHaveTextContent( /a-document.pdf/i );
```
We can probably simplify by going with a string instead of a regular expression here.

```suggestion
		expect( link ).toHaveTextContent( 'a-document.pdf' );
```

tyxla reviewed (approved):
Looks and works great, thanks once again @brookewp  🚀

brookewp committed:
Simplify regex to a string

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

getdave commented:
Thank you for all the effort here @brookewp 👏

t-hamano commented:
#57391 was reported in relation to this PR.

The current truncation logic relies on media always having an extension. If there is no extension, the URL without the protocol will be returned. So, as reported in #57391, if the media is rendered server-side and the URL does not include the file extension, truncation does not seem to occur correctly.

The logic of filterURLForDisplay may need to be improved a bit.
